pkg/webhook/authorizer/withselectors.go

```diff
@@ -0,0 +1,122 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package authorizer
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/go-logr/logr"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/clock"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	versionutils "github.com/gardener/gardener/pkg/utils/version"
+)
+
+// WithSelectorsChecker checks whether the 'AuthorizeWithSelectors' feature is enabled in the kube-apiserver.
+// TODO(rfranzke): Remove this interface once the lowest supported Kubernetes version is 1.34.
+type WithSelectorsChecker interface {
+	// IsPossible returns true if the 'AuthorizeWithSelectors' feature is enabled in the kube-apiserver.
+	IsPossible() (bool, error)
+}
+
+// NewWithSelectorsChecker creates a new WithSelectorsChecker.
+func NewWithSelectorsChecker(ctx context.Context, log logr.Logger, clientSet kubernetes.Interface, clock clock.Clock) WithSelectorsChecker {
+	return &withSelectorsChecker{
+		ctx:       ctx,
+		log:       log.WithName("authorize-with-selectors-checker"),
+		clientSet: clientSet,
+		clock:     clock,
+
+		isPossible:    false,
+		nextCheckTime: ptr.To(clock.Now()),
+	}
+}
+
+type withSelectorsChecker struct {
+	ctx       context.Context
+	log       logr.Logger
+	clientSet kubernetes.Interface
+	clock     clock.Clock
+
+	isPossible    bool
+	nextCheckTime *time.Time
+}
+
+func (w *withSelectorsChecker) IsPossible() (bool, error) {
+	if w.nextCheckTime != nil && !w.clock.Now().UTC().Before(w.nextCheckTime.UTC()) {
+		enabled, mustCheckAgain, err := w.isAuthorizeWithSelectorsFeatureEnabled()
+		if err != nil {
+			return false, fmt.Errorf("failed checking whether 'AuthorizeWithSelectors' feature is enabled: %w", err)
+		}
+
+		w.isPossible = enabled
+		if !mustCheckAgain {
+			w.nextCheckTime = nil
+			w.log.Info("No need not check again")
+		} else {
+			w.nextCheckTime = ptr.To(w.clock.Now().UTC().Add(10 * time.Minute))
+			w.log.Info("Must check again, caching result", "nextCheckTime", w.nextCheckTime.String())
+		}
```

Comment:
I am in favour of less negation. Hence, I propose to switch the if/else block:

```suggestion
		if mustCheckAgain {
			w.nextCheckTime = ptr.To(w.clock.Now().UTC().Add(10 * time.Minute))
			w.log.Info("Must check again, caching result", "nextCheckTime", w.nextCheckTime.String())
		} else {
			w.nextCheckTime = nil
			w.log.Info("No need not check again")
		}
```